cql3/statements/select_statement.cc

```diff
@@ -1144,6 +1147,38 @@ query::partition_slice indexed_table_select_statement::get_partition_slice_for_g
             }
 
             partition_slice_builder.with_ranges(clustering_restrictions->bounds_ranges(options));
+        } else if (has_token(_restrictions->get_partition_key_restrictions()->expression) && is_token_v2_index) {
+            auto token_ranges = _restrictions->get_partition_key_restrictions()->bounds_ranges(options);
+            std::vector<query::clustering_range> all_bounds;
+
+            for (auto& token_range : token_ranges) {
+                auto& start_ring_postition = token_range.start()->value();
+                auto& end_ring_position = token_range.end()->value();
+
+                const bool include_start = start_ring_postition.bound() == dht::ring_position::token_bound::start;
+                const bool include_end = end_ring_position.bound() == dht::ring_position::token_bound::end;
+
+                int64_t start_position = start_ring_postition.token().raw();
+                int64_t end_position = end_ring_position.token().raw();
+
+                auto clustering_restrictions = ::make_shared<restrictions::single_column_clustering_key_restrictions>(_view_schema, false);
+
+                auto min_token_expr = expr::binary_operator{
+                    &token_cdef, include_start ? expr::oper_t::GTE : expr::oper_t::GT,
+                    ::make_shared<cql3::constants::value>(cql3::raw_value::make_value(long_type->decompose(start_position)))};
+                auto max_token_expr = expr::binary_operator{
+                    &token_cdef, include_end ? expr::oper_t::LTE : expr::oper_t::LT,
+                    ::make_shared<cql3::constants::value>(cql3::raw_value::make_value(long_type->decompose(end_position)))};
+
+                auto token_restriction = ::make_shared<restrictions::single_column_restriction>(token_cdef);
+                token_restriction->expression = expr::make_conjunction(std::move(min_token_expr), std::move(max_token_expr));
+                clustering_restrictions->merge_with(token_restriction);
```

Comment:
Can these be built during the prepare phase?

Reply:
(using an unbound term for the token values)

Reply:
/cc @dekimir for advice

Reply:
Yes, this can theoretically be done in a `prepare` method, with abstract_marker for rhs in min/max_token_expr.